src/audio/xm64.c

```diff
@@ -0,0 +1,240 @@
+	return xm_get_number_of_channels(player->ctx);
+}
+
+void xm64player_play(xm64player_t *player, int first_ch) {
+	assert(first_ch + xm_get_number_of_channels(player->ctx) <= MIXER_MAX_CHANNELS);
+
+	if (!player->playing) {
+		// XM64 header contains the optimal size for sample buffers on each
+		// channel, to minimize memory consumption. To configure it, bump
+		// the frequency of each channel to an unreasonably high value (we don't
+		// know how much we need, so shoot high), but then limit the buffer size
+		// to the optimal value.
+		for (int i=0; i<player->ctx->module.num_channels; i++) {
+			// If the value is 0, the channel is not used. We don't have a way
+			// to convey this (0 would be interpreted as "no limit"), so just
+			// avoid calling the limit function altogether.
+			if (player->ctx->ctx_size_stream_sample_buf[i] != 0)
+				mixer_ch_set_limits(first_ch+i, 0, 1e9, player->ctx->ctx_size_stream_sample_buf[i]);
+		}
+
+		mixer_add_event(0, tick, player);
+		player->first_ch = first_ch;
+		player->playing = true;
+	}
+}
+
+void xm64player_stop(xm64player_t *player) {
+	// Let the mixer callback stop playing
+	player->playing = false;
+}
+
+void xm64player_tell(xm64player_t *player, int *patidx, int *row, float *secs) {
+	// Disable interrupts to try to avoid race conditions with the player
+	// running in a different thread. Notice that this is not sufficient
+	// (you would need some kind of mutex), but let's say good enough,
+	// especially since the audio thread is going to be higher priority.
+	disable_interrupts();
+	uint8_t patidx_, row_; uint64_t samples;
+	xm_get_position(player->ctx, &patidx_, NULL, &row_, &samples);
+	bool has_pending_seek = player->seek.patidx >= 0;
+	if (patidx) *patidx = has_pending_seek ? player->seek.patidx : patidx_;
+	if (row) *row = has_pending_seek ? player->seek.row : row_;
+	if (secs) *secs = (float)samples / (float)player->ctx->rate;
+	enable_interrupts();
+}
+
+void xm64player_seek(xm64player_t *player, int patidx, int row, int tick) {
+	// Schedule seeking at next tick. Don't seek here to avoid
+	// race conditions when the mixer is running in another thread.
+	disable_interrupts();
+	player->seek.patidx = patidx;
+	player->seek.row = row;
+	player->seek.tick = tick;
+	enable_interrupts();
+}
+
+void xm64player_set_vol(xm64player_t *player, float volume) {
+	// Store the volume in the libxm context as amplification.
+	// 0.25f is the default suggested value, so we scale by it.
+	player->ctx->amplification = volume * 0.25f;
+}
+
+void xm64player_close(xm64player_t *player) {
+	// FIXME: we need to stop playing without racing with the audio thread.
+	// This is not correct and may crash.
+	disable_interrupts();
+	if (player->playing) {
+		mixer_remove_event(tick, player);
+		player->playing = false;
+	}
+	for (int i=0;i<player->ctx->module.num_channels;i++) {
+		mixer_ch_stop(player->first_ch+i);
+		mixer_ch_set_limits(player->first_ch, 0, 0, 0);
```

Comment:
`mixer_ch_stop` just stops the channel (stops the audio). 

`mixer_ch_set_limits` is a permanent configuration of the channel that gives the mixer information on how big to allocate the samplebuffer. By default, the mixer has to be conservative and allocate buffers for each channel that are big enough for the worst case (eg: 16-bit 44.1Khz playback). The xmplayer uses this API to minimize memory consumption (the converter calculates the *exact* size required for playback on each channel).

The limits are not set for each waveform being played because it would require too many deallocations/allocations. Instead, we calculate limits which are good for the whole song and stick with them. Experimentally, this is a good enough solution, as most XM songs require < 100Kb of RDRAM total (samplebuffers plus XM data structures) so I would say it's good enough.

Here we're shutting down the player, so, to be a good citizen, the player removes any limit that have been configured in the mixer, to leave it in pristine state. Otherwise, somebody else might try to play something else on one of the channels that were previous used by the xmplayer, and fail to playback because it would find a too small buffer.